src/core/namespace.c

```diff
@@ -875,16 +875,55 @@ static bool namespace_info_mount_apivfs(const char *root_directory, const NameSp
         assert(ns_info);
 
         /*
-         * ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=,
-         * since to protect the API VFS mounts, they need to be around in the
-         * first place... and RootDirectory= or RootImage= need to be set.
+         * If one of ProtectControlGroups=, ProtectKernelTunables= or
+         * PrivateDevices= is set then MountAPIVFS= is implied, since to
+         * protect the API VFS mounts, they need to be around in the first
+         * place... and RootDirectory= or RootImage= need to be set.
          */
 
         /* root_directory should point to a mount point */
         return root_directory &&
                 (ns_info->mount_apivfs ||
                  ns_info->protect_control_groups ||
-                 ns_info->protect_kernel_tunables);
+                 ns_info->protect_kernel_tunables ||
+                 ns_info->private_dev);
+}
+
+static int append_mount_apivfs(MountEntry **p, const char *root_directory,
+                               const NameSpaceInfo *ns_info, bool ignore_protect) {
+        int r = 0;
+
+        assert(p);
+
+        if (!root_directory)
+                return 0;
+
+        /*
+         * If one of ProtectControlGroups=, ProtectKernelTunables= or PrivateDevices=
+         * are set, then mount the API file systems as they are in a protect
+         * mode.
+         */
+        if (ns_info->protect_control_groups ||
+            ns_info->protect_kernel_tunables || ns_info->private_dev) {
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/sys",
+                        .mode = SYSFS,
+                        .ignore = ignore_protect,
+                };
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/proc",
+                        .mode = PROCFS,
+                        .ignore = ignore_protect,
+                };
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/dev",
+                        .mode = PRIVATE_DEV,
+                        .ignore = ignore_protect,
+                };
+        } else if (ns_info->mount_apivfs)
+                r = append_static_mounts(p, apivfs_table, ELEMENTSOF(apivfs_table), ignore_protect);
```

Comment:
hmm, not sure i follow, but what is the effective difference between the two if blocks? it appears to me append_static_mounts() with the apivfs_table does pretty much the same as the open-coded three MountEntry additions you do above?